modules/context/repo.go

```diff
@@ -523,12 +523,9 @@ func getRefName(ctx *Context, pathType RepoRefType) string {
 
 // URL to redirect to for deprecated URL scheme
 func repoRefRedirect(ctx *Context) string {
-	urlPath := ctx.Req.URL.String()
-	idx := strings.LastIndex(urlPath, ctx.Params("*"))
-	if idx < 0 {
-		idx = len(urlPath)
-	}
-	return path.Join(urlPath[:idx], ctx.Repo.BranchNameSubURL())
+	urlPath := strings.TrimSuffix(ctx.Req.URL.String(), ctx.Params("*"))
+	urlPath = strings.TrimPrefix(urlPath, "/")
+	return setting.AppURL + path.Join(urlPath, ctx.Repo.BranchNameSubURL())
```

Comment:
nit: could we use [`URLJoin`](https://github.com/go-gitea/gitea/blob/52e11b24bf5e395d83ea58c1b0fd6922efe16add/modules/markup/html.go#L116) instead of manually assembling the URL (it may make sense to also move `URLJoin` to `modules/util`)

Reply:
@ethantkoenig @lafriks Do you want me to do this change in this PR? It's a lot of lines changed, would it be in the scope of this PR?

Reply:
I think moving into path.Join would be enough for this PR

Reply:
@lafriks You can't move `setting.AppURL` to `path.Join`. Or you mean `markup.URLJoin`?

Reply:
I mean changing from `return setting.AppURL + path.Join(urlPath, ctx...` to `return path.Join(setting.AppURL, urlPath, ctx....`

Reply:
It's imposible because `path.Join` strips `http://example.com` to `http:/example.com`

Reply:
You can use that suburl variable then so than?

Reply:
@lafriks I can use `setting.AppSubURL`.